### PR TITLE
[Foundation] Add hack enabling factory intializer

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?

I'm not sure this can be accepted or not.
But this is convenient in some cases.

Using that:
- Simplify copy constructor of `NSSet` and `NSDictionary`
- Eliminate `getVaList` from `init(format: ..)` initalizer of `NSPredicate`, `NSExpression` and `NSString`

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by rintaro** [comment]
@dabrahams, @gribozavr 
How do you think about this?

**Comment by gribozavr** [comment]
I'm concerned about ABI stability issues.  We are essentially exploiting a compiler bug with this technique.  We'd want to replace this trick with a proper factory initializer, but we'd have to do this before ABI stability.

**Comment by dabrahams** [comment]
@rintaro I discovered that technique at WWDC2015 but I thought we had 'fixed' the compiler so it no longer worked!  I love the amount of code deletion.

@gribozavr So add a `FIXME(ABI)` and be done?

**Comment by gribozavr** [comment]
> @gribozavr So add a FIXME(ABI) and be done?

Not a bad idea.

**Comment by rintaro** [comment]
@gribozavr, @dabrahams, Thank you for the comment!
Assigning `self` in initializer in protocol extension is used in stdlib `Integers.swift.gyb`.
Just to make sure, adopting such protocols by class types may cause some problems in the future?

**Comment by gribozavr** [comment]
I am concerned that the need to fix the compiler bug will arise before we get proper factory initializers.  If that would be the case, we would be left without any options to implement these APIs.  Same concerns apply to `Integers.swift.gyb`.

File: stdlib/public/SDK/Foundation/Foundation.swift [hunk 1]
```diff
@@ -14,6 +14,15 @@
 import CoreFoundation
 import CoreGraphics
 
+// HACK: Enable factory initializer, using  protocol extension that allows
+// assign to `self`.
+protocol _Factory {}
```
Comment (slavapestov) [inline stdlib/public/SDK/Foundation/Foundation.swift]: Oh that's cute :)

**Comment by slavapestov** [comment]
I don't think assignment to self in a protocol extension initializer is a bug. Now we do want to change the ABI to make factory initializers more efficient, but the language feature is fine.

**Comment by jrose-apple** [comment]
Yeah, it's assignment to self in mutating methods that's problematic.

**Comment by tkremenek** [comment]
This was opened in 2016.  Is this still relevant?

**Comment by rintaro** [comment]
We should revisit here when https://forums.swift.org/t/allow-self-x-in-class-convenience-initializers/15924 is implemented.
CC: @jckarter @slavapestov @phausler 

We cannot simply adapt this PR to the current code layout, because `type(of: self)` in closure are no longer allowed.
```swift
    self.init(_factory: {
      type(of: self).init() // error: 'self' used before 'self.init' call
    })
```
Also, I found a couple of bugs which makes it difficult to adapt this PR to the current compiler.
https://bugs.swift.org/browse/SR-9429
https://bugs.swift.org/browse/SR-9430

Closing for now.

**Pull request closed without merging**